### PR TITLE
Issue 104700: Clarify error message for collection property with no getter

### LINKED ISSUE #104700: Misleading error message when trying to deserialize JSON into a collection property without a getter
### Description

When trying to deserialize a JSON string into a class containing a property with a collection type (`Dictionary` or `IEnumerable`) serializer throws an exception which says about a property missing a setter despite it being present.

### Reproduction Steps

```csharp
JsonSerializer.Deserialize<Test>("{\"Data\":[1]}");

public class Test
{
    [JsonRequired]
    public List<int> Data
    {
        set => FirstValue = value[0];
    }
    
    [JsonIgnore]
    public int FirstValue { get; private set; }
}
```

### Expected behavior

Either:
- JSON is deserialized correctly;
- Error message says exactly what the problem is and doesn’t mislead a user

### Actual behavior

The following exception is thrown:
```
System.InvalidOperationException: JsonPropertyInfo 'Data' defined in type 'Test' is marked required but does not specify a setter.
   at System.Text.Json.ThrowHelper.ThrowInvalidOperationException_JsonPropertyRequiredAndNotDeserializable(JsonPropertyInfo jsonPropertyInfo)
   at System.Text.Json.Serialization.Metadata.JsonPropertyInfo.Configure()
   at System.Text.Json.Serialization.Metadata.JsonTypeInfo.ConfigureProperties()
   at System.Text.Json.Serialization.Metadata.JsonTypeInfo.Configure()
   at System.Text.Json.Serialization.Metadata.JsonTypeInfo.<EnsureConfigured>g__ConfigureSynchronized|172_0()
   at System.Text.Json.JsonSerializerOptions.GetTypeInfoInternal(Type type, Boolean ensureConfigured, Nullable`1 ensureNotNull, Boolean resolveIfMutable, Boolean fallBackToNearestAncestorType)
   at Program.<Main>$(String[] args)
```

### Regression?

No, this behavior is present at least since .NET 7 when required properties were introduced.

### Known Workarounds

Adding a getter to the property, it may return null or just outright throw an exception.

### Configuration

.NET 8.0.6 running on macOS 14.5 ARM64; issue is not specific to the configuration.

### Other information

Similar to (although different scenarios): https://github.com/dotnet/runtime/issues/92330

### PR DESCRIPTION
**Pull request by julien98Qc**
- Added verification to distinguish between different error types.
- Renamed existing test to more accurately reflect that it tests required properties without a setter.
- Added a new test to ensure that a collection property without a getter throws an InvalidOperationException.

Fixes #104700

### CODE REVIEW

**Comment by julien98Qc** [comment]
> @julien98Qc please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @dotnet-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > 
> > ```
> > @dotnet-policy-service agree
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> > 
> > ```
> > @dotnet-policy-service agree company="Microsoft"
> > ```
> 
> Contributor License Agreement

@dotnet-policy-service agree

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonPropertyInfo.cs [hunk 1]
```diff
@@ -448,19 +448,7 @@ internal void Configure()
 
             if (IsRequired)
             {
-                if (!CanDeserialize &&
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonPropertyInfo.cs:451]: Could you keep the logic inlined please? Extracting into a separate method makes the diff harder to review and IMHO it shouldn't make a big difference from a readability perspective. Thanks.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonPropertyInfo.cs [hunk 2]
```diff
@@ -472,6 +460,31 @@ internal void Configure()
...
         internal abstract void DetermineReflectionPropertyAccessors(MemberInfo memberInfo, bool useNonPublicAccessors);
 
+        private void EnsureRequiredPropertyIsDeserializable()
+        {
+            if (!CanDeserialize)
+            {
+                bool isCollectionProperty = (EffectiveConverter.ConverterStrategy & (ConverterStrategy.Enumerable | ConverterStrategy.Dictionary)) != 0;
+                bool isRequiredParameterRespected = AssociatedParameter?.IsRequiredParameter is true && Options.RespectRequiredConstructorParameters;
+
+                if (isCollectionProperty)
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonPropertyInfo.cs:470]: Is this really only improving the error message or is it expanding failures for more types? It seems to me that if we have a collection property and `RespectRequiredConstructorParameters` is set to false this is failing when it wasn't before this change.

File: src/libraries/System.Text.Json/tests/Common/RequiredKeywordTests.cs [hunk 3]
```diff
@@ -564,7 +564,7 @@ public async Task ChangingPropertiesWithRequiredKeywordToNotBeRequiredAllowsDese
         }
 
         [Fact]
-        public async Task RequiredNonDeserializablePropertyThrows()
+        public async Task RequiredPropertyWithoutSetterThrows()
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/tests/Common/RequiredKeywordTests.cs:567]: Please don't rename existing unit tests as it breaks historical tracking in CI.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonPropertyInfo.cs [hunk 4]
```diff
@@ -472,6 +460,31 @@ internal void Configure()
         [RequiresDynamicCode(JsonSerializer.SerializationRequiresDynamicCodeMessage)]
         internal abstract void DetermineReflectionPropertyAccessors(MemberInfo memberInfo, bool useNonPublicAccessors);
 
+        private void EnsureRequiredPropertyIsDeserializable()
+        {
+            if (!CanDeserialize)
+            {
+                bool isCollectionProperty = (EffectiveConverter.ConverterStrategy & (ConverterStrategy.Enumerable | ConverterStrategy.Dictionary)) != 0;
+                bool isRequiredParameterRespected = AssociatedParameter?.IsRequiredParameter is true && Options.RespectRequiredConstructorParameters;
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonPropertyInfo.cs:468]: Nit: this expression is being evaluated even for branches that won't use it.

File: src/libraries/System.Text.Json/tests/Common/RequiredKeywordTests.cs [hunk 5]
```diff
@@ -610,6 +610,23 @@ public class ClassWithRequiredExtensionDataProperty
...
+        [Fact]
+        public async Task RequiredCollectionPropertyWithoutGetterThrows()
+        {
+            string json = """{"Foo":"foo","Bar":"bar"}""";
+            InvalidOperationException exception = await Assert.ThrowsAsync<InvalidOperationException>(
+                async () => await Serializer.DeserializeWrapper<ClassWithRequiredCollectionPropertyWithoutGetter>(json));
+            Assert.Contains(nameof(ClassWithRequiredCollectionPropertyWithoutGetter.TestCollectionPropertyWithoutGetter), exception.Message);
+        }
+
+        public class ClassWithRequiredCollectionPropertyWithoutGetter
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/tests/Common/RequiredKeywordTests.cs:622]: Just to make sure we're not regressing things here, could you also add a test for class like so:
```C#
class ClassWithCollectionCtorParam(Dictionary<string, JsonElement> value)
{
    public Dictionary<string, JsonElement> Value { get; } = value;
}
```
And then try to deserialize it using an `options` instance with `RespectRequiredConstructorParameters` set to `true`.

File: src/libraries/System.Text.Json/tests/Common/RequiredKeywordTests.cs [hunk 6]
```diff
@@ -610,6 +610,23 @@ public class ClassWithRequiredExtensionDataProperty
             public required Dictionary<string, JsonElement>? TestExtensionData { get; set; }
         }
 
+        [Fact]
+        public async Task RequiredCollectionPropertyWithoutGetterThrows()
+        {
+            string json = """{"Foo":"foo","Bar":"bar"}""";
+            InvalidOperationException exception = await Assert.ThrowsAsync<InvalidOperationException>(
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/tests/Common/RequiredKeywordTests.cs:617]: I don't think this is the correct way to address the problem. A fix should instead try to make the scenario work as expected. If you remove `JsonRequired` from the repro in #104700 then deserialization works as expected and this should too.

**Review by eiriktsarpalis — CHANGES_REQUESTED** [review]
Per https://github.com/dotnet/runtime/pull/104861/files#r1692841347 I don't believe clarifying the error message is the correct way to address the issue. We should instead make the impact scenario work similar to how it does when `[JsonRequired]` isn't being specified.

**Comment by eiriktsarpalis** [comment]
@julien98Qc Thank you for taking the time to contribute a fix!

**Comment by julien98Qc** [comment]
@eiriktsarpalis Thank you for reviewing :D

**Pull request closed without merging**